src/wp-includes/default-filters.php

```diff
@@ -192,6 +192,7 @@
 add_filter( 'the_title', 'convert_chars' );
 add_filter( 'the_title', 'trim' );
 
+add_filter( 'the_content', 'apply_block_hooks_to_content', 8 ); // BEFORE do_blocks().
```

Comment:
`apply_block_hooks_to_content` parses blocks, traverses the block tree (to insert hooked blocks), and reserializes them. Note that it's followed immediately by [`do_blocks`](https://developer.wordpress.org/reference/functions/do_blocks/) (on the line below), which _also_ parses blocks.

This means that in theory, there's some potential for optimization (parse only once inside of `do_blocks`, and traverse to insert hooked blocks there).

In practice, I'm not sure if that's viable, or if that might cause some collisions; in particular if `do_blocks` is already used in other contexts where hooked blocks are already inserted (through a different mechanism), e.g. in templates or `wp_navigation` posts.